keps/sig-auth/2862-fine-grained-kubelet-authz/README.md

```diff
@@ -0,0 +1,1003 @@
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+To determine if the caller has the required permissions for a particular request
+kubelet creates a `SubjectAccessReview` request to `kube-apiserver`. The
+`SubjectAccessReviews` are currently populated with:
+
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  user: user1
+  uid: 64167384-10aa-4361-bcef-526ab51d1e4d
+  groups:
+  - groups1
+  resourceAttributes:
+    group: ""
+    version: "v1"
+    resource: "nodes"
+    namespace: ""
+    name: "node-1"
+    subresource: "proxy"
+    verb: "GET"
+```
+
+The following information is passed into the `SubjectAccessReview` request
+- Requesting user (generally from a `TokenReviewRequest` or client certificate authentication)
+- Request verb
+- A resource request with:
+  - APIGroup: ""
+  - APIVersion: "v1"
+  - Resource: "nodes"
+  - Name: nodeName (the name of the node the request was against)
+  - Subresource: which can be one of: proxy, log, metrics, spec, stats
+
+The subresource is determined by the path of the request. `kube-apiserver` upon
+receiving this request will check if the user in the `SubjectAccessReview` has
+the RBAC permissions. The RBAC for the sameple request above would look like as
+follows:-
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: nodes-proxy
+rules:
+- apiGroups: [""]
+  resources: ["nodes/proxy"]
+  verbs: ["get", "create"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: nodes-proxy-global
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: User
+  name: user1
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: nodes-proxy
+```
+
+We will add a new feature-gate `KubeletFineGrainedAuthz` that will be defaulted to
+`false` in `alpha`. When this feature-gate is set to `true` kubelet will first send
+a `SubjectAccessReview` using the fine-grained verb for `healthz` and `pods` endpoints.
+If that request fails kubelet will retry with the coarse-grained verb (`proxy`).
+
+When `kube-apiserver` communicates with the kubelet
+its user is bound to the `system:kubelet-api-admin` role. When the feature-gate 
+is enabled we will update the `ClusterRole` as follow:-
+
+```diff
+  apiVersion: rbac.authorization.k8s.io/v1
+  kind: ClusterRole
+  metadata:
+    annotations:
+      rbac.authorization.kubernetes.io/autoupdate: "true"
+    labels:
+      kubernetes.io/bootstrapping: rbac-defaults
+    name: system:kubelet-api-admin
+  rules:
+  - apiGroups:
+    - ""
+    resources:
+    - nodes
+    verbs:
+    - get
+    - list
+    - watch
+  - apiGroups:
+    - ""
+    resources:
+    - nodes
+    verbs:
+    - proxy
```

Comment:
This seems to be allowing full access to proxy to all nodes. Is that what we want to demonstrate?

Reply:
This role already exists and has full access to proxy to all nodes. I just want to demonstrate the additional permissions we will add to this role.